### PR TITLE
Eliminate nullness and rawtype errors from java SDK Read class

### PR DESCRIPTION
**Pull request by kennknowles**
This eliminates outright nullness and rawtype errors where possible, and localizes some cases that are reasonable compromises or hacks that I don't want to mess with.

One small step towards #20507

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by damondouglas** [comment]
Run Java PreCommit

**Review by damondouglas — CHANGES_REQUESTED** [review]
I cloned `kennknowles:read-nullness` and at `a41ac61a43`, I ran:
```sh
./gradlew :javaPrecommit
```

and discovered:

```
> Task :sdks:java:core:test

org.apache.beam.sdk.util.UnboundedScheduledExecutorServiceTest > testThreadsAreAddedOnlyAsNeededWithContention FAILED
    java.lang.AssertionError at UnboundedScheduledExecutorServiceTest.java:556
```

Running, `./gradlew :sdks:java:core:test` produced the same result.

Yet, running `./gradlew :sdks:java:core:test` at `4183e747be` did not have any such error.

It got late before I can analyze more why.

**Comment by kennknowles** [comment]
I'll take a look. I actually just ensured it compiled before pushing out, since I wasn't intending to make functional changes.

**Comment by kennknowles** [comment]
huh, interestingly I failed to repro the failure

**Comment by kennknowles** [comment]
This is undoubtedly something I added because the thing looked like it was required to be non-null.

```
Caused by: java.lang.IllegalStateException
	at org.apache.beam.sdk.util.Preconditions.checkStateNotNull(Preconditions.java:452)
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java [hunk 1]
```diff
@@ -839,16 +850,16 @@ public CheckpointMark getCheckpointMark() {
...
+        checkStateNotNull(restrictionCoder);
         // For caching reader, we don't care about the watermark.
         return restrictionCoder.structuralValue(
             UnboundedSourceRestriction.create(
                 source, checkpoint, BoundedWindow.TIMESTAMP_MIN_VALUE));
       }
 
       private void initializeCurrentReader() throws IOException {
-        Preconditions.checkState(currentReader == null);
+        checkStateNotNull(currentReader);
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java:862]: oops

**Comment by kennknowles** [comment]
Run Java PreCommit

**Comment by kennknowles** [comment]
Run RAT PreCommit

**Comment by kennknowles** [comment]
Run Java_IOs_Direct PreCommit

**Comment by kennknowles** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by kennknowles** [comment]
Run Java_Kinesis_IO_Direct PreCommit

**Review by damondouglas — APPROVED** [review]
LGTM :-)

**Comment by damondouglas** [comment]
FYI Run Java_GCP_IO_Direct PreCommit shows success even though it has a yellow light.

**Pull request merged**